src/Structure/Editor/EditorBlock.php

```diff
@@ -23,63 +37,365 @@
+
+		return strtolower( str_replace( DIRECTORY_SEPARATOR, '-', $name_parts[1] ) );
+	}
+
+	/**
+	 * Get the path values of the calling class as an indexed array.
+	 *
+	 * Returns an indexed array which contains the path type (e.g., plugins or themes) and
+	 * the path name, which is the directory name for the plugin or theme.
+	 *
+	 * @author Jeremy Ward <jeremy.ward@webdevstudios.com>
+	 * @since  2019-12-27
+	 * @throws ReflectionException Within ReflectionClass.
+	 * @throws Exception If unexpected directory name in wp-content.
+	 */
+	private function derive_path_values() {
+		$relative_path = str_replace( WP_CONTENT_DIR, '', $this->get_file_path() );
+		$parts         = explode( DIRECTORY_SEPARATOR, ltrim( $relative_path, DIRECTORY_SEPARATOR ) );
+
+		$this->set_wpcontent_dirname( $parts[0] ?? '' );
+		$this->package_dirname = $parts[1] ?? '';
+	}
+
+	/**
+	 * Get the file path if one is provided, or derive one if not.
+	 *
+	 * @author Jeremy Ward <jeremy.ward@webdevstudios.com>
+	 * @since  2019-12-27
+	 * @throws ReflectionException Within ReflectionClass.
+	 * @return false|string
+	 */
+	public function get_file_path() {
+		return $this->file_path ?: ( new \ReflectionClass( get_called_class() ) )->getFileName();
 	}
 
 	/**
 	 * Get the URL to an editor block asset.
 	 *
+	 * This method checks for the full path to the block assets and checks whether the provided asset
+	 * can be read at the given path. If it is,
+	 *
 	 * @param string $asset The asset for which to retrieve the URL.
 	 *
 	 * @author Jeremy Ward <jeremy.ward@webdevstudios.com>
 	 * @since  2019-02-17
 	 * @return string
-	 * @throws \Exception If the class does not exist or the asset is not readable.
+	 * @throws Exception If the class does not exist or the asset is not readable.
 	 */
 	public function get_asset_url( string $asset ) {
-		if ( ! $this->file_path ) {
-			$this->file_path = str_replace( ABSPATH, '', dirname( ( new \ReflectionClass( static::class ) )->getFileName(), 2 ) );
+		$asset_path = $this->get_assets_path( $asset );
+
+		if ( ! is_readable( $asset_path ) ) {
+			throw new Exception( "Could not find requested asset at {$asset_path}." );
 		}
 
-		$asset_path = ABSPATH . $this->file_path . '/assets/' . $asset;
+		return 'plugins' === $this->wpcontent_dirname
+			? plugin_dir_url( $this->file_path . $asset ) . $asset
+			: trailingslashit( get_stylesheet_directory_uri() ) . $this->file_path . $asset;
+	}
 
-		if ( ! is_readable( $asset_path ) ) {
-			throw new \Exception( "Could not find requested asset at {$asset_path}." );
+	/**
+	 * Get the root file path to this block's assets.
+	 *
+	 * @param string $asset The asset to look up.
+	 * @throws Exception If the file path value is not set or if the asset cannot be found.
+	 *
+	 * @author Jeremy Ward <jeremy.ward@webdevstudios.com>
+	 * @since  2019-08-02
+	 * @return string
+	 */
+	protected function get_assets_path( string $asset ) : string {
+		if ( 'plugins' === $this->wpcontent_dirname ) {
```

Comment:
I think we could simplify this class by moving this logic out of here. I think it would be an improvement if this class not know (and not care) what type of project it is being used in.

Reply:
Before we make changes here, can you propose an alternative?

I agree that the class shouldn't have to care about what kind of project it's being used in, but it does need to have some kind of meaningful way to locate its own JavaScript and CSS assets for the purposes of registering itself.

In this iteration, if you create a new `EditorBlock` object, you are expected to provide it with a valid name at a minimum. This class will use that name to derive everything else it needs to be able to register a block for you (save you creating the JavaScript and CSS assets yourself).

I could see how a `PathLocator` class could perhaps be shipped alongside the package which the `EditorBlock` uses. Is that what you're proposing?

Reply:
To put it another way, the `EditorBlock` model has JS and CSS assets as a core part of itself. To register itself, it needs to know where they are. I'm fine extracting that logic into another class under the same namespace, but the ultimate goal should be that an engineer has to extend only this class when registering a new block. Whatever that looks like, I'm open to recommendations.

Reply:
I'm not sure what a refactoring would look like. When I think about registering a block type in an OOP way, I imagine something like

```
$editor_script_asset_obj = new Asset( 
    'myblock-block-script',
    '/scripts/my-block.js', // Path to asset relative to this file (or maybe some other standard)
    $dependencies // string array of dependencies
)

$block_config = new BlockConfig( 'webdevstudios/myblock` );
$block_config->set_editor_script( $editor_script_asset_obj );
$block_config->set_editor_style( $editor_style_asset_obj );
$block_config->set_style( $style_asset_obj );
$block_config->set_render_callback( [ $my_object, 'my_render_callback' ] );
$block_config->set_attributes( [] ); // A default could be set to skip this line.

$block_type_registrar->register( $block_config );
```

Clearly, this involves adding a lot of new classes and refactoring and I haven't completely thought this through. The problem I see currently is this `EditorBlock` class is doing too much and we would benefit by breaking some of these parts out.